### ISSUE TITLE
Support forward-declaration of well-known types?

### ISSUE THREAD

**Issue by nelfin** [nelfin/pylint-protobuf#45]
> ```
> E5901: Field 'GetCurrentTime' does not appear in the declared fields of protobuf-generated class 'Timestamp' and will raise AttributeError on access (protobuf-undefined-attribute)
> ```
> Has the fix made it to the release?
> 
> Or maybe I confused the detection by including the Timestamp definition verbatim in my proto-file. I've done this to simplify the lives of my users since the ubuntu protoc package does not include WKTs definitions.
> 
> ```proto
> // well known types not available on ubuntu protoc system install
> // import "google/protobuf/timestamp.proto";
> message google {
>     message protobuf {
>         message Timestamp {
>           int64 seconds = 1;
>           int32 nanos = 2;
>         }
>     }
> }
> ```
> This is enough so that this `Timestamp` is recognized as WKT.

_Originally posted by @zwn in https://github.com/nelfin/pylint-protobuf/issues/14#issuecomment-808870851_

**Comment by nelfin** [comment]
Frankly, I'm surprised this works at all

**Comment by nelfin** [comment]
It looks like this is why:

https://github.com/protocolbuffers/protobuf/blob/87dd07b4367b7676af42105d0d102f4e536c248b/python/google/protobuf/internal/python_message.py#L151-L152

When you declare a nested message like above, or if you'd done `package google.protobuf;` and `message Timestamp` then the "full name" of that type is `google.protobuf.Timestamp` (the file does not make a difference, since I presume you haven't declared a package). See this excerpt from the generated code:

```python
_GOOGLE_PROTOBUF_TIMESTAMP = _descriptor.Descriptor(
  name='Timestamp',
  full_name='google.protobuf.Timestamp',
  filename=None,
  file=DESCRIPTOR,
# ...
```

I'll have a look at this and see what it takes to match that behaviour, but I'm not sure this'll be an easy fix. If so, it might be a while before I get to it.

**Comment by nelfin** [comment]
This didn't turn out to be as hard as I thought. I still don't like how wacky it is but whatever. @zwn, can you try the latest changes in `pip install git+https://github.com/nelfin/pylint-protobuf@89b01399c8048db60d415b661fd94a52f458debc` and see if they address all of your use-cases?

**Comment by zwn** [comment]
@nelfin Well, this was fast! I can confirm it works for my use case. :+1:  Thank you!

> I still don't like how wacky it is but whatever.

wrt wacky - it follows my mental model of how protobuf works. In my world the "import" is like textual include with the difference that the result is put inside a namespace inferred from the path of the imported file. So it does not feel wacky to me but I totally understand that under a different mental model of the behavior it is a no-go.

**Comment by nelfin** [comment]
> it follows my mental model of how protobuf works. In my world the "import" is like textual include with the difference that the result is put inside a namespace inferred from the path of the imported file. So it does not feel wacky to me but I totally understand that under a different mental model of the behavior it is a no-go.

Fair enough! Thanks for helping me understand your mental model.

**Comment by nelfin** [comment]
This has been released in `v0.20.0` just now
